src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -0,0 +1,509 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.minecolonies.entity.ai.util.AIState.*;
+
+/**
+ * Abstract class which contains all the guard basics let it be range, melee or magic.
+ */
+public abstract class AbstractEntityAIGuard extends AbstractEntityAISkill<JobGuard>
+{
+    /**
+     * The start search distance of the guard to track/attack entities may get more depending on the level.
+     */
+    private static final double MAX_ATTACK_DISTANCE = 20.0D;
+
+    /**
+     * Distance the guard starts searching.
+     */
+    private static final int START_SEARCH_DISTANCE = 5;
+
+    /**
+     * Basic delay after operations.
+     */
+    private static final int BASE_DELAY = 1;
+
+    /**
+     * Max amount the guard can shoot arrows before restocking.
+     */
+    private static final int MAX_ATTACKS = 50;
+
+    /**
+     * Y range in which the guard detects other entities.
+     */
+    private static final double HEIGHT_DETECTION_RANGE = 10D;
+
+    /**
+     * Path that close to the patrol target.
+     */
+    private static final int PATH_CLOSE      = 3;
+
+    /**
+     * Worker gets this distance times building level away from his building to patrol.
+     */
+    private static final int PATROL_DISTANCE = 20;
+
+    /**
+     * Horizontal range in which the guard picks up items
+     */
+    private static final double RANGE_HORIZONTAL_PICKUP = 20.0D;
+    /**
+     * Vertical range in which the guard picks up items
+     */
+    private static final double RANGE_VERTICAL_PICKUP   = 2.0D;
+
+    /**
+     * Amount of ticks after which the guard stops trying to gather and tries to get onto another task.
+     */
+    private static final int   STUCK_WAIT_TICKS        = 20;
+
+    /**
+     * The amount of time to wait while walking to items
+     */
+    private static final int   WAIT_WHILE_WALKING      = 5;
+
+    /**
+     * The range in which the guard picks up items.
+     */
+    private static final int ITEM_PICKUP_RANGE = 3;
+
+    /**
+     * Checks if the guard should dump its inventory.
+     */
+    private static final int DUMP_AFTER_ACTIONS = 10;
+
+    /**
+     * The current target.
+     */
+    protected EntityLivingBase targetEntity;
+    /**
+     * Amount of arrows already shot or sword hits dealt.
+     */
+    protected int attacksExecuted = 0;
+    /**
+     * The distance the guard is searching entities in currently.
+     */
+    private int currentSearchDistance = START_SEARCH_DISTANCE;
+    /**
+     * Current goTo task.
+     */
+    private BlockPos currentPatrolTarget;
+    /**
+     * Containing all close entities.
+     */
+    private List<Entity> entityList;
+
+    /**
+     * Positions of all items that have to be collected.
+     */
+    @Nullable
+    private List<BlockPos> items;
+
+    /**
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)